zkevm-circuits/src/pi_circuit.rs

```diff
@@ -0,0 +1,978 @@
+        region.assign_advice(|| "number", self.raw_public_inputs, offset, || Ok(number))?;
+        raw_pi_vals[offset] = number;
+        offset += 1;
+
+        // timestamp
+        let timestamp = rlc(block_values.timestamp.to_le_bytes(), randomness);
+        region.assign_advice(|| "timestamp", self.block_value, offset, || Ok(timestamp))?;
+        region.assign_advice(
+            || "timestamp",
+            self.raw_public_inputs,
+            offset,
+            || Ok(timestamp),
+        )?;
+        raw_pi_vals[offset] = timestamp;
+        offset += 1;
+
+        // difficulty
+        let difficulty = rlc(block_values.difficulty.to_le_bytes(), randomness);
+        region.assign_advice(|| "difficulty", self.block_value, offset, || Ok(difficulty))?;
+        region.assign_advice(
+            || "difficulty",
+            self.raw_public_inputs,
+            offset,
+            || Ok(difficulty),
+        )?;
+        raw_pi_vals[offset] = difficulty;
+        offset += 1;
+
+        // base_fee
+        let base_fee = rlc(block_values.base_fee.to_le_bytes(), randomness);
+        region.assign_advice(|| "base_fee", self.block_value, offset, || Ok(base_fee))?;
+        region.assign_advice(
+            || "base_fee",
+            self.raw_public_inputs,
+            offset,
+            || Ok(base_fee),
+        )?;
+        raw_pi_vals[offset] = base_fee;
+        offset += 1;
+
+        // chain_id
+        let chain_id = rlc(block_values.chain_id.to_le_bytes(), randomness);
+        region.assign_advice(|| "chain_id", self.block_value, offset, || Ok(chain_id))?;
+        let chain_id_cell = region.assign_advice(
+            || "chain_id",
+            self.raw_public_inputs,
+            offset,
+            || Ok(chain_id),
+        )?;
+        raw_pi_vals[offset] = chain_id;
+        offset += 1;
+
+        for prev_hash in block_values.history_hashes {
+            let prev_hash = rlc(prev_hash.to_le_bytes(), randomness);
+            region.assign_advice(|| "prev_hash", self.block_value, offset, || Ok(prev_hash))?;
+            region.assign_advice(
+                || "prev_hash",
+                self.raw_public_inputs,
+                offset,
+                || Ok(prev_hash),
+            )?;
+            raw_pi_vals[offset] = prev_hash;
+            offset += 1;
+        }
+
+        Ok(chain_id_cell)
+    }
+
+    /// Assigns the extra fields (not in block or tx tables):
+    ///   - state root
+    ///   - parent block hash
+    /// to the raw_public_inputs column and stores a copy in a
+    /// vector for computing RLC(raw_public_inputs).
+    #[allow(clippy::type_complexity)]
+    fn assign_extra_fields(
+        &self,
+        region: &mut Region<'_, F>,
+        extra: ExtraValues,
+        randomness: F,
+        raw_pi_vals: &mut Vec<F>,
+    ) -> Result<(AssignedCell<F, F>, AssignedCell<F, F>), Error> {
+        let mut offset = BLOCK_LEN;
+        // block hash
```

Comment:
NOTE: In the future we will need to verify the block hash in the circuits; and when that day arrives, this will be useful :)